File: Sugeno/Fuzzy.cpp
#include "Fuzzy.h"
#include <Arduino.h>


// ------------------- Konstruktor -------------------

Fuzzy::Fuzzy() {
    LANE = 500;
    LINE = 3000;
    CENTER = 1570;
    left_dir = false;
}

// ------------------- inference -------------------

Directions Fuzzy::setDirection(int smallLeft, 
                    int bigLeft,
                    int smallRight,
                    int bigRight)
{
   Directions d;
   float l = (smallLeft + bigLeft) * 0.5;
   float r = (smallRight + bigRight) * 0.5;

   if (l > r) left_dir = true; 
   else left_dir = false;

   if (left_dir) {
    d.small = smallLeft;
    d.big = bigLeft;
   }
   else{
    d.small = smallRight;
    d.big = bigRight;
   }

   return d;
}

InputMembershipDegrees Fuzzy::inference(const Directions &d)
{
    InputMembershipDegrees m;
    m.outerLine  = 0.0f;
    m.innerLine    = 0.0f;

    // Auswertung innerer Sensor
    if (d.small <= LANE) m.innerLine = 0;
    else if (LANE < d.small && d.small < LINE){
        m.innerLine = float(d.small - LANE) / float(LINE - LANE);
    }
    else {
        m.innerLine = 1;
    }

    // Auswertung äußerer Sensor
    if (d.big <= LANE) m.outerLine = 0;
    else if (LANE < d.big && d.big < LINE){
        m.outerLine = float(d.big - LANE) / float(LINE - LANE);
    }
    else{
        m.outerLine = 1;
    }


    return m;
}

// ------------------- rules -------------------
//
// Entspricht der Python-Methode:
// def rules(self, membershipdegrees):
//   if smallLeft or bigLeft -> right
//   if smallRight or bigRight -> left
//   if normal -> normal
//

OutputMembershipDegrees Fuzzy::rules(const InputMembershipDegrees &m) {
    OutputMembershipDegrees o;
    o.rule1   = 0.0f;
    o.rule2 = 0.0f;
    o.rule3 = 0.0f;
    o.rule4  = 0.0f;

    // Regel 1 : keiner der Sensoren berührt die Linie
    o.rule1 = (1.0 - m.outerLine) * (1.0 - m.innerLine);
    // Regel 2: Nur der äußere Sensor berührt die Linie
    o.rule2 = (1.0 - m.outerLine) * m.innerLine;
    // Regel 3: beide Sensoren berühren die Linie
    o.rule3 = m.outerLine * m.innerLine;
    // Regel 4: nur der innere Sensor berührt die Linie
    o.rule4 = m.outerLine * (1.0 - m.innerLine);


    return o;
}

// ------------------- defuzzify -------------------

float Fuzzy::defuzzify(const OutputMembershipDegrees &o) { //"&o" übergibt das Objekt keine Kopie
    float steering_pwm = CENTER;

    float noSteering = 0.0;
    float lightSteering = 60.0;
    float mediumSteering = 160.0;
    float maxSteering = 200.0;

    float z = (o.rule1 * noSteering + o.rule2 * lightSteering + o.rule3 * mediumSteering + o.rule4 * maxSteering)/(o.rule1+o.rule2+o.rule3+o.rule4);

    if (left_dir){
        steering_pwm = CENTER + z;
    }
    else {
        steering_pwm = CENTER - z;
    }

    return steering_pwm;
}

File: Sugeno/fuzzy.py
import matplotlib.pyplot as plt

class Fuzzy:
    def __init__(self):
        self.LANE = 500
        self.LINE = 3000
        self.CENTER = 1570
        self.left_dir = False

    def setDirection(self, sL, bL, sR, bR):
        ''' Takes Sensorvalues small Left, big Left, small Right, big Right. \n
        Returns: small, big (from the side of the bigger deviation) '''
        l = (sL + bL) * 0.5
        r = (sR + bR) * 0.5

        # Richtung bestimmen
        self.left_dir = True if l > r else False

        if self.left_dir:
            s = sL
            b = bL
        else:
            s = sR
            b = bR

        return s, b

    def inference(self, s, b):
        '''  Array [boolean leftDirection, float line membershipdegree outer Sensor,
        float line membershipdegree inner Sensor] '''


        # Zugehörigkeitsgrade Linie bestimmen
        if s <= self.LANE: line_i = 0
        elif self.LANE < s < self.LINE:
            line_i = (s - self.LANE) / (self.LINE - self.LANE)
            pass
        else:
            line_i = 1

        if b <= self.LANE: line_o = 0
        elif self.LANE < b < self.LINE:
            line_o = (b - self.LANE) / (self.LINE - self.LANE)
            pass
        else:
            line_o = 1


        return line_o, line_i

    def ruleSet(self, line_i, line_o):
        # keiner der Sesnoren berührt die Linie:
        w1 = (1-line_i)*(1-line_o)
        # Nur der äußere Sensor berührt die Linie
        # if not line_i and line_o:
        w2 = (1-line_i)*line_o
        # beide Sensoren berühren die Linie
        # if line_i and line_0:
        w3 = line_i*line_o
        # nur der innere Sensor berührt die Linie
        w4 = line_i*(1-line_o)

        return w1, w2, w3, w4

    def calculateOutput(self,w1, w2, w3, w4):
        no_steering = 0
        light_steering = 60
        medium_steering = 160
        max_steering = 200

        z = (w1 * no_steering + w2 * light_steering + w3 * medium_steering + w4 * max_steering)/(w1+w2+w3+w4)

        if self.left_dir:
            output = self.CENTER + z
        else:
            output = self.CENTER - z


        return output

fuzzy = Fuzzy()
s,b = fuzzy.setDirection(0,0,3350,100)
line_o, line_i = fuzzy.inference(s,b)
w1,w2,w3,w4 = fuzzy.ruleSet(line_o, line_i)
print(fuzzy.calculateOutput(w1,w2,w3,w4))

# fuzzy = Fuzzy()
# outputL = []
# outputR = []
# sR = []
# bR = []
# sL = []
# bL = []
#
#
# for i in range(0, 4095, 50):
#     for j in range(0,4095, 50):
#         sL.append(i)
#         bL.append(j)
#
#         s, b = fuzzy.setDirection(i, j, 0, 0)
#         line_o, line_i = fuzzy.inference(s, b)
#         w1,w2, w3, w4 = fuzzy.ruleSet(line_o, line_i)
#         outputR.append(fuzzy.calculateOutput(w1, w2, w3, w4))
#
# for i in range(0, 4095, 50):
#     for j in range(0,4095, 50):
#         sR.append(i)
#         bR.append(j)
#
#         s, b = fuzzy.setDirection(0, 0, i, j)
#         line_o, line_i = fuzzy.inference(s, b)
#         w1,w2, w3, w4 = fuzzy.ruleSet(line_o, line_i)
#         outputL.append(fuzzy.calculateOutput(w1, w2, w3, w4))
#
#
# fig = plt.figure()
# ax = fig.add_subplot(111, projection='3d')
#
# ax.scatter(sL,bL,outputR, s=10, color='red')
# ax.scatter(sR,bR,outputL, s=10, color='blue')
# ax.scatter(3350,100, 1370, s=10, color='green')
#
# ax.set_xlabel('small')
# ax.set_ylabel('big')
# ax.set_zlabel('pwm')
#
# plt.show()
#
#
#
#
